### DISCUSSION TITLE
OpenSSL 3.0.8 - How to fallback to default provider when property fips=yes set and FIPS provider is loaded

### DISCUSSION THREAD

**Discussion by Phani-Sreenivasa-Prasad** [openssl/openssl#24740] [Q&A]
As part of FIPS compliance for one of our project, , we are explicitely loading the FIPS provider programatically along with default provider(enabled by default in conf file) and re-initializing the OpenSSL library.  This is because we have few applications which doesnt need to use FIPS provider but only default provider hence we load both.
Along with loading the FIPS provider we also set the fips=yes property using EVP_set_default_property API. 

Our goal is that all crypto should go to FIPS provider  when FIPS is enabled and we are achieving that with fips=yes set.  But at the same time, usage of non-fips compliant algorithms (for eg: PKCS12_parse won't work with FIPS provider. This is just one example) we want those to fallback to Default provider.   Ofcourse, we are trying to define local property string to override provider=default/ -fips option and using EVP_XXX_fetch for explicitely loading algo from "Default" provider.  

But the FIPS standard would evolve and tomorrow it may stop supporting SHA-1 for eg.,and we have to folllow the same approach of defining local property string. Again it has to go through code changes everywhere wherever SHA-1 is used to explictely load it from Default.

Instead, is there a way we can always fallback to Default provider when it is already loaded and  if FIPS provider rejects an algorithm? 1. Basically some priority order for look up of algorithms with each loaded provider?
2.  or is it possible to explictely map a specific algorithms to load always from a specific providers programatically or in the openssl.cnf sections?

**Comment by nhorman** [comment]
You can use the Prefrence prefix operator for this:

https://www.openssl.org/docs/man3.0/man7/property.html

in your property query, you can specify "fips ?= yes" to prefer fips, but fall back to another provider if fips isn't available

**Comment by mattcaswell** [comment]
> "fips ?= yes" 

That should be "?fips=yes"

**Reply by Phani-Sreenivasa-Prasad** [reply]
can we use this "?fips=yes" for default_properties as well? I mean something like below.

while calling **_EVP_set_default_properties(NULL, "?fips=yes");_**  


or
[algorithm_sect]
default_properties = "?fips=yes"

**Reply by mattcaswell** [reply]
Yes

**Comment by mattcaswell** [comment]
The problem with this approach is of course your're never quite sure whether in any given situation you are FIPS compliant or not. Another issues is that it may not solve all your FIPS problems. E.g. an algorithm might be present in the FIPS provider but have restrictions on its use imposed on it in the future (e.g. a signature algorithm might be restricted so that it can only be used for verifying signatures but not creating new ones).

One thing to consider using is library contexts. In one library context you loads the fips/base providers, and in another one you load the default library context. You then must make sure you pass the correct libctx in any given situation. This won't "automatically" fall back though - you would still need to make a conscious choice. But I'm not really convinced that automatically falling back is really a very good idea. Normally you want to definitively know whether or not you are FIPS compliant (otherwise what's the point?)

**Reply by Phani-Sreenivasa-Prasad** [reply]
I understood.  But we don't claim FIPS compliance for all the applications. and we need ensure applications that are NOT needed to be fips compliant to fallback to default provider and doenst break.  For eg: if an end customer is using PKCS12 based certificates with our app and since PKCS12KDF is not supported by FIPS provider, we would like to fallback to Default provider for PKCS12_create and PKCS12_parse API.
